packages/python/test/unit.test.ts

```diff
@@ -0,0 +1,78 @@
+
+let warningMessages: string[];
+const originalConsoleWarn = console.warn;
+beforeEach(() => {
+  warningMessages = [];
+  console.warn = m => {
+    warningMessages.push(m);
+  };
+});
+
+afterEach(() => {
+  console.warn = originalConsoleWarn;
+});
+
+it('should only match supported versions, otherwise throw an error', async () => {
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: '3.9' })
+  ).toHaveProperty('runtime', 'python3.9');
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: '3.6' })
+  ).toHaveProperty('runtime', 'python3.6');
+});
+
+it('should ignore minor version in vercel dev', async () => {
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: '3.9', isDev: true })
+  ).toHaveProperty('runtime', 'python3');
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: '3.6', isDev: true })
+  ).toHaveProperty('runtime', 'python3');
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: '999', isDev: true })
+  ).toHaveProperty('runtime', 'python3');
+  expect(warningMessages).toStrictEqual([]);
+});
+
+it('should select latest version when no Piplock detected', async () => {
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: undefined })
+  ).toHaveProperty('runtime', 'python3.9');
+  expect(warningMessages).toStrictEqual([]);
+});
+
+it('should select latest version and warn when invalid Piplock detected', async () => {
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: '999' })
+  ).toHaveProperty('runtime', 'python3.9');
+  expect(warningMessages).toStrictEqual([
+    'Warning: Python version "999" detected in Pipfile.lock is invalid and will be ignored. http://vercel.link/python-version',
+  ]);
+});
+
+it('should throw for discontinued versions', async () => {
+  const realDateNow = Date.now.bind(global.Date);
+  global.Date.now = () => new Date('2022-07-31').getTime();
+  expect(() =>
+    getSupportedPythonVersion({ pipLockPythonVersion: '3.6' })
+  ).toThrow(
+    'Python version "3.6" detected in Pipfile.lock is discontinued and must be upgraded. This change is the result of a decision made by an upstream infrastructure provider (AWS).'
+  );
+  expect(warningMessages).toStrictEqual([]);
+  global.Date.now = realDateNow;
+});
+
+it('should warn for deprecated versions, soon to be discontinued', async () => {
+  const realDateNow = Date.now.bind(global.Date);
+  global.Date.now = () => new Date('2021-07-01').getTime();
+
+  expect(
+    getSupportedPythonVersion({ pipLockPythonVersion: '3.6' })
+  ).toHaveProperty('runtime', 'python3.6');
+  expect(warningMessages).toStrictEqual([
+    'Error: Python version "3.6" detected in Pipfile.lock is deprecated. Deployments created on or after 2022-07-18 will fail to build. This change is the result of a decision made by an upstream infrastructure provider (AWS). http://vercel.link/python-version',
+  ]);
+
+  global.Date.now = realDateNow;
```

Comment:
Can we wrap this test body in a `try-finally` so that an exception wouldn't cause test pollution in future tests? Same goes for the other tests that manipulate `global.Date.now`.

```suggestion
  const realDateNow = Date.now.bind(global.Date);
  try {
    global.Date.now = () => new Date('2021-07-01').getTime();

    expect(
      getSupportedPythonVersion({ pipLockPythonVersion: '3.6' })
    ).toHaveProperty('runtime', 'python3.6');
    expect(warningMessages).toStrictEqual([
      'Error: Python version "3.6" detected in Pipfile.lock is deprecated. Deployments created on or after 2022-07-18 will fail to build. This change is the result of a decision made by an upstream infrastructure provider (AWS). http://vercel.link/python-version',
    ]);
  } finally {
    global.Date.now = realDateNow;
  }
```

Reply:
I updated the code to use `beforeEach()` and `afterEach()` to avoid putting try/finally in every test.